packages/lib/services/e2ee/crypto.ts

```diff
@@ -0,0 +1,132 @@
+			cipher = createCipheriv(algorithm, key, iv, { authTagLength: authTagLength } as CipherGCMOptions) as CipherGCM;
+			iv = iv || nodeRandomBytes(12); // "For IVs, it is recommended that implementations restrict support to the length of 96 bits, to promote interoperability, efficiency, and simplicity of design." - NIST SP 800-38D
+		} else {
+			throw new Error(_('Unknown cipher algorithm: %s', algorithm));
+		}
+
+		cipher.setAAD(associatedData, { plaintextLength: Buffer.byteLength(data) });
+
+		const encryptedData = Buffer.concat([cipher.update(data), cipher.final()]);
+		const authTag = cipher.getAuthTag();
+
+		return Buffer.concat([encryptedData, authTag]);
+	},
+
+	decryptRaw: (data: CryptoBuffer, algorithm: CipherAlgorithm, key: CryptoBuffer, iv: CryptoBuffer, authTagLength: number, associatedData: CryptoBuffer | null) => {
+		if (associatedData === null) {
+			associatedData = Buffer.alloc(0);
+		}
+
+		let decipher = null;
+		if (algorithm === CipherAlgorithm.AES_256_GCM || algorithm === CipherAlgorithm.AES_192_GCM || algorithm === CipherAlgorithm.AES_128_GCM) {
+			decipher = createDecipheriv(algorithm, key, iv, { authTagLength: authTagLength } as CipherGCMOptions) as DecipherGCM;
+		} else {
+			throw new Error(_('Unknown decipher algorithm: %s', algorithm));
+		}
+
+		const authTag = data.subarray(-authTagLength);
+		const encryptedData = data.subarray(0, data.byteLength - authTag.byteLength);
+		decipher.setAuthTag(authTag);
+		decipher.setAAD(associatedData, { plaintextLength: Buffer.byteLength(data) });
+
+		let decryptedData = null;
+		try {
+			decryptedData = Buffer.concat([decipher.update(encryptedData), decipher.final()]);
+		} catch (error) {
+			throw new Error(`Authentication failed! ${error}`);
+		}
+
+		return decryptedData;
+	},
+
+	encrypt: async (password: string, iterationCount: number, salt: CryptoBuffer | null, data: CryptoBuffer) => {
+
+		// default encryption parameters
+		const cipherAlgorithm = CipherAlgorithm.AES_256_GCM;
+		const authTagLength = 16; // 128 bits
+		const digest = Digest.sha512;
+		const keySize = 32; // For CipherAlgorithm.AES_256_GCM, 256 bits -> 32 bytes
+
+		// default encryption parameters won't appear in result
+		const result: EncryptionResult = {
+			iter: iterationCount,
+			salt: '',
+			iv: '',
+			ct: '', // cipherText
+		};
+		salt = salt || await crypto.randomBytes(32); // 256 bits
+		const iv = await crypto.randomBytes(12); // 96 bits
+
+		const key = await crypto.pbkdf2Raw(password, salt, iterationCount, keySize, digest);
+		const encrypted = crypto.encryptRaw(data, cipherAlgorithm, key, iv, authTagLength, null);
+
+		result.salt = salt.toString('base64');
+		result.iv = iv.toString('base64');
+		result.ct = encrypted.toString('base64');
+
+		return result;
+	},
+
+	decrypt: async (password: string, data: EncryptionResult) => {
+
+		// default encryption parameters
+		const cipherAlgorithm = data.algo || CipherAlgorithm.AES_256_GCM;
```

Comment:
I think I mentioned it before but there should be no optional parameter - everything should be explicitly set by the caller.

Reply:
I've added a new interface `EncryptionParameters` which contains all parameters.